File: backend/Services/Parsers/BalatroParser.cs
using System.Globalization;
using System.Text.RegularExpressions;
using backend.Models;

namespace backend.Services.Parsers
{
    public class BalatroParser : ResultParser
    {
        private ILogger<BalatroParser> _logger;

        public BalatroParser(ILogger<BalatroParser> logger) : base(logger)
        {
            _logger = logger;
        }

        public override bool CountWinner => true;
        public override string GameName => "Balatro Daily Challenge";
        public override bool GolfScoring => false;
        public override string? HelpText
        {
            get
            {
                var today = DateTime.Today;
                int seed = today.Year * 10000 + today.Month * 100 + today.Day;
                var rand = new Random(seed);

                var decks = new string[]
                {
                    "Red",
                    "Blue",
                    "Yellow",
                    "Green",
                    "Black",
                    "Magic",
                    "Nebula",
                    "Ghost",
                    "Abandoned",
                    "Checkered",
                    "Zodiac",
                    "Painted",
                    "Anaglyph",
                    "Plasma",
                    "Erratic"
                };

                var stakes = new string[]
                {
                    "White",
                    "Red",
                    "Green",
                    "Black",
                    "Blue",
                    "Purple",
                    "Orange",
                    "Gold"
                };

                var randomDeck = decks[rand.Next(0, decks.Length)];
                var randomStake = stakes[rand.Next(0, stakes.Length)];

                return $"Seed: today's random word.\nDeck: {randomDeck}.\nStake: {randomStake}.\nEntry examples: \"b 9 16 60000\", \"B a 9 r 16 bh 60000\", or \"Balatro ante 9 round 16 best hand 60,000\".";
            }
        }
        private readonly string AnteGroup = "Ante";
        private readonly string RoundGroup = "Round";
        private readonly string BestHandGroup = "BestHand";

        protected override Regex Parser => new Regex($"(balatro|b) (a|ante)? ?(?<{AnteGroup}>\\d+) (r|round)? ?(?<{RoundGroup}>\\d+) (bh|best hand)? ?(?<{BestHandGroup}>[\\d,.]+)", RegexOptions.IgnoreCase);
        public override string Url => "https://www.playbalatro.com/";

        protected override string GetCleanResult(string result, Match parserResults)
        {
            if (!parserResults.Groups.ContainsKey(AnteGroup) || !parserResults.Groups.ContainsKey(RoundGroup) || !parserResults.Groups.ContainsKey(BestHandGroup))
            {
                return "Balatro";
            }

            var ante = parserResults.Groups[AnteGroup].Value ?? "";

            var round = parserResults.Groups[RoundGroup].Value ?? "";

            var bestHand = parserResults.Groups[BestHandGroup].Value ?? "";

            return "Balatro" + "\n" + "Ante: " + ante + "\n" + "Round: " + round + "\n" + "Best hand: " + bestHand;
        }

        public override string? GetScoreValue(DailyResult dailyResult)
        {
            if (dailyResult.Scores == null)
            {
                return null;
            }

            return string.Join(", ", dailyResult.Scores);
        }

        protected override DailyResult SetScore(DailyResult dailyResult, Match parserResults)
        {
            if (!parserResults.Groups.ContainsKey(AnteGroup) || !parserResults.Groups.ContainsKey(RoundGroup) || !parserResults.Groups.ContainsKey(BestHandGroup))
            {
                return dailyResult;
            }

            var anteResult = parserResults.Groups[AnteGroup].Value;
            var ante = int.Parse(anteResult, NumberStyles.AllowThousands, CultureInfo.InvariantCulture);

            var roundResult = parserResults.Groups[RoundGroup].Value;
            var round = int.Parse(roundResult, NumberStyles.AllowThousands, CultureInfo.InvariantCulture);

            var bestHandResult = parserResults.Groups[BestHandGroup].Value
                .Replace(",", "")
                .Replace(".", "");
            var bestHand = int.Parse(bestHandResult, NumberStyles.AllowThousands, CultureInfo.InvariantCulture);

            dailyResult.Scores = [
                ante,
                round,
                bestHand
            ];

            return dailyResult;
        }
    }
}

File: lambdas/src/parsers/balatro-parser.ts
import { DailyResult } from '../models/daily-result';
import { ResultParser, getGroup, hasGroup } from './base/result-parser';
import { getNowEasternStandardTime } from '../utils/time-utility';

export class BalatroParser extends ResultParser {
  readonly countWinner = true;
  readonly gameName = 'Balatro Daily Challenge';
  readonly golfScoring = false;
  readonly url = 'https://www.playbalatro.com/';
  
  private readonly randomDeck: string;
  private readonly randomStake: string;
  
  constructor() {
    super();
    
    const nowEt = getNowEasternStandardTime();
    const seed = nowEt.getFullYear() * 10000 + (nowEt.getMonth() + 1) * 100 + nowEt.getDate();
    
    // Note: Uses different RNG algorithm than C# backend (LCG vs Knuth subtractive)
    const rand = this.seededRandom(seed);
    
    const decks = [
      'Red', 'Blue', 'Yellow', 'Green', 'Black', 'Magic', 'Nebula',
      'Ghost', 'Abandoned', 'Checkered', 'Zodiac', 'Painted',
      'Anaglyph', 'Plasma', 'Erratic'
    ];
    
    const stakes = [
      'White', 'Red', 'Green', 'Black', 'Blue', 'Purple', 'Orange', 'Gold'
    ];
    
    this.randomDeck = decks[Math.floor(rand() * decks.length)];
    this.randomStake = stakes[Math.floor(rand() * stakes.length)];
  }
  
  private seededRandom(seed: number): () => number {
    return () => {
      seed = (seed * 9301 + 49297) % 233280;
      return seed / 233280;
    };
  }
  
  get helpText(): string {
    return `Seed: today's random word.\nDeck: ${this.randomDeck}.\nStake: ${this.randomStake}.\nEntry examples: "b 9 16 60000", "B a 9 r 16 bh 60000", or "Balatro ante 9 round 16 best hand 60,000".`;
  }
  
  protected readonly parser = new RegExp(
    `(balatro|b) (a|ante)? ?(?<ante>\\d+) (r|round)? ?(?<round>\\d+) (bh|best hand)? ?(?<bestHand>[\\d,.]+)`,
    'i'
  );
  
  protected getCleanResult(result: string, match: RegExpMatchArray): string {
    if (!hasGroup(match, 'ante') || !hasGroup(match, 'round') || !hasGroup(match, 'bestHand')) {
      return 'Balatro';
    }
    
    const ante = getGroup(match, 'ante') || '';
    const round = getGroup(match, 'round') || '';
    const bestHand = getGroup(match, 'bestHand') || '';
    
    return `Balatro\nAnte: ${ante}\nRound: ${round}\nBest hand: ${bestHand}`;
  }
  
  getScoreValue(dailyResult: DailyResult): string | null {
    if (!dailyResult.scores) return null;
    return dailyResult.scores.join(', ');
  }
  
  protected setScore(dailyResult: DailyResult, match: RegExpMatchArray): DailyResult {
    if (!hasGroup(match, 'ante') || !hasGroup(match, 'round') || !hasGroup(match, 'bestHand')) {
      return dailyResult;
    }
    
    // Strip thousands separators (commas) to match C#'s NumberStyles.AllowThousands behavior
    const anteStr = (getGroup(match, 'ante') || '0').replace(/,/g, '');
    const roundStr = (getGroup(match, 'round') || '0').replace(/,/g, '');
    // bestHand strips both commas and periods (for different locale formats)
    const bestHandStr = (getGroup(match, 'bestHand') || '0').replace(/[,.]/g, '');
    
    dailyResult.scores = [
      parseInt(anteStr, 10),
      parseInt(roundStr, 10),
      parseInt(bestHandStr, 10)
    ];
    
    return dailyResult;
  }
}
